### ISSUE TITLE
2024 No migration for a particular RPIT scenario (async?)

### ISSUE THREAD

**Issue by ehuss** [rust-lang/rust#132819]
Sorry, I'm not certain how to describe this scenario or why it doesn't seem to work. This could also probably be minimized more.

I tried this code:

```rust
#![warn(rust_2024_compatibility)]
use std::pin::Pin;

pub struct AttachContainerResults {
    pub input: Pin<Box<dyn AsyncWrite + Send>>,
}

pub trait AsyncWrite {}
struct S;
impl AsyncWrite for S {}

pub async fn attach_container(this: &i32) -> AttachContainerResults {
    let write = process_upgraded(this).await;

    AttachContainerResults {
        input: Box::pin(write),
    }
}

async fn process_upgraded(_this: &i32) -> impl AsyncWrite {
    S
}
```

I expected to see this happen: Should generate a warning about it not working in 2024. At least, I cannot correlate the error with the scenarios described in the [migration](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/rpit-lifetime-capture.html#migration) docs that describe scenarios that are not supported.

Instead, this happened: This does not generate any warnings about it not working in 2024. After migrating to 2024, you get an error:

```
error: lifetime may not live long enough
  --> src/main.rs:16:16
   |
12 | pub async fn attach_container(this: &i32) -> AttachContainerResults {
   |                                     - let's call the lifetime of this reference `'1`
...
16 |         input: Box::pin(write),
   |                ^^^^^^^^^^^^^^^ coercion requires that `'1` must outlive `'static`
```

I believe the solution here is to add:

```diff
     }
 }
 
-async fn process_upgraded(_this: &i32) -> impl AsyncWrite {
+async fn process_upgraded(_this: &i32) -> impl AsyncWrite + use<> {
     S
 }
```

From messing around, it seems like the `async` part of `process_upgraded` is the key factor that is preventing the `impl_trait_overcaptures` lint from firing.

### Meta

`rustc --version --verbose`:
```
rustc 1.84.0-nightly (b91a3a056 2024-11-07)
binary: rustc
commit-hash: b91a3a05609a46f73d23e0995ae7ebb4a4f429a5
commit-date: 2024-11-07
host: aarch64-apple-darwin
release: 1.84.0-nightly
LLVM version: 19.1.3
```

**Comment by compiler-errors** [comment]
Isn't this just a duplicate of https://github.com/rust-lang/rust/issues/132809#issuecomment-2466346725 ? This should be fixed by https://github.com/rust-lang/rust/pull/132817.

**Comment by ehuss** [comment]
Oh, indeed it probably is! I didn't make the connection since the errors were different.

**Comment by compiler-errors** [comment]
The error messages will result in a wide array of different borrowck errors that all have to do with things being required to live longer. Only some of them are intercepted by #131186, but due to the nature of NLL, we won't be able to detect all of them.
